enterprise_subsidy/apps/api/v1/views.py

```diff
@@ -1 +1,180 @@
+    role_assignment_class = EnterpriseSubsidyRoleAssignment
+
+    def get_permission_required(self):
+        """
+        Return permissions required for the current requested action.
+
+        We override the function here instead of just setting the ``permission_required`` class attribute because that
+        attribute only supports requiring a single permission for the entire viewset.  This override logic allows for
+        the permission required to be based conditionally on the type of action.
+        """
+        permission_for_action = {
+            # Note: right now all actions require the same permission, but I'll leave this complexity in here in
+            # anticipation that other write actions will be added soon.
+            "list": PERMISSION_CAN_READ_SUBSIDIES,
+            "retrieve": PERMISSION_CAN_READ_SUBSIDIES,
+        }
+        permission_required = permission_for_action.get(self.request_action, PERMISSION_NOT_GRANTED)
+        return permission_required
+
+    @property
+    def requested_enterprise_customer_uuid(self):
+        """
+        Look in the query parameters for an enterprise customer UUID.
+        """
+        return utils.get_enterprise_uuid_from_request_query_params(self.request)
+
+    @property
+    def requested_subsidy_uuid(self):
+        """
+        Fetch the subsidy UUID from the URL location.
+
+        For detail endpoints, the PK can simply be found in ``self.kwargs``.
+        """
+        return self.kwargs.get('uuid')
+
+    @property
+    def base_queryset(self):
+        """
+        Required by the ``PermissionRequiredForListingMixin``.
+        For non-list actions, this is what's returned by ``get_queryset()``.
+        For list actions, some non-strict subset of this is what's returned by ``get_queryset()``.
+        """
+        kwargs = {}
+        if self.requested_enterprise_customer_uuid:
+            kwargs.update({'enterprise_customer_uuid': self.requested_enterprise_customer_uuid})
+        if self.requested_subsidy_uuid:
+            kwargs.update({'uuid': self.requested_subsidy_uuid})
+
+        return Subsidy.objects.filter(**kwargs).prefetch_related(
+            # Fields used for calculating the ledger balance.
+            'ledger__transactions__state',
+            'ledger__transactions__quantity',
+            'ledger__transactions__reversal',
+            'ledger__transactions__reversal__state',
+            'ledger__transactions__reversal__quantity',
+        ).order_by('uuid')
+
+
+class TransactionViewSet(
+    PermissionRequiredForListingMixin,
+    mixins.ListModelMixin,
+    mixins.RetrieveModelMixin,
+    mixins.CreateModelMixin,
+    viewsets.GenericViewSet,
+):
+    """
+    ViewSet for the Transaction model.
+
+    Currently, this only supports listing, retrieving, creating, and reversing transactions::
+
+      GET  /api/v1/transactions/
+      GET  /api/v1/transactions/{transaction_uuid}/
+      POST /api/v1/transactions/?subsidy_uuid={subsidy_uuid}
```

Comment:
to be more restful, `subsidy_uuid` would probably be a body/post param rather than a query param

Reply:
For reads: it feels more natural, and perhaps necessary, to list transactions for a particular subsidy.
We might want to split out a viewset for reading transactions into something like `TransactionReadViewset`, which can use `PermissionRequiredForListingMixin`, and have the URL path be scoped to the subsidy, i.e. `GET /api/v1/subsidies/[subsidy_uuid/transactions`, which feels more natural to me.  

Then something like `TransactionWriteViewset` to support creation and reversals.  This might make it easier to support more natural paths for creation and reversal, i.e. `POST /api/v1/subsidies/{subsidy_uuid}/transactions`  to create a transaction (although maybe we don't necessarily need the subsidy uuid as part of the path for reversing a transaction, since you'd have the `transaction_uuid`, which is the PK of the thing to reverse).

Maybe some of these endpoints can be plain old views that use the `@permission_required` rbac decorator?